lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -774,74 +824,62 @@ static bool canReplaceCopiedSelf(FullApplySite Apply,
 /// FIXME: Protocol methods (witness or default) that return Self will be given
 /// a new return type. This implementation fails to update the type signature of
 /// SSA uses in those cases. Currently we bail out on methods that return Self.
-SILInstruction *
-SILCombiner::createApplyWithConcreteType(FullApplySite Apply,
-                                         const ConcreteExistentialInfo &CEI,
-                                         SILBuilderContext &BuilderCtx) {
-  assert(Apply.getOrigCalleeType()->isPolymorphic());
+SILInstruction *SILCombiner::createApplyWithConcreteType(
+    FullApplySite Apply,
+    llvm::SmallDenseMap<unsigned, const ConcreteExistentialInfo *, 8> &CEIs,
+    SILBuilderContext &BuilderCtx) {
 
-  // Don't specialize apply instructions that return the callee's Self type,
-  // because this optimization does not know how to substitute types in the
-  // users of this apply. In the function type substitution below, all
-  // references to OpenedArchetype will be substituted. So walk to type to find
-  // all possible references, such as returning Optional<Self>.
-  if (Apply.getType().getASTType().findIf(
-          [&CEI](Type t) -> bool { return t->isEqual(CEI.OpenedArchetype); })) {
-    return nullptr;
-  }
-  // Bail out if any non-self arguments or indirect result that refer to the
-  // OpenedArchetype. The following optimization substitutes all occurrences of
-  // OpenedArchetype in the function signature, but will only rewrite the self
-  // operand.
-  //
-  // Note that the language does not allow Self to occur in contravariant
-  // position. However, SIL does allow this and it can happen as a result of
-  // upstream transformations. Since this is bail-out logic, it must handle all
-  // verifiable SIL.
-  for (auto Arg : Apply.getArgumentsWithoutSelf()) {
-    if (Arg->getType().getASTType().findIf([&CEI](Type t) -> bool {
-          return t->isEqual(CEI.OpenedArchetype);
-        })) {
-      return nullptr;
-    }
-  }
-  // The apply can only be rewritten in terms of the concrete value if it is
-  // legal to pass that value as the self argument.
-  if (CEI.isCopied && (!CEI.InitExistential ||
-                       !canReplaceCopiedSelf(Apply, CEI.InitExistential, DA))) {
-    return nullptr;
-  }
+  // Ensure that the callee is polymorphic.
+  assert(Apply.getOrigCalleeType()->isPolymorphic());
 
-  // Create a set of arguments.
+  // Create the new set of arguments to apply including their substitutions.
+  SubstitutionMap NewCallSubs = Apply.getSubstitutionMap();
   SmallVector<SILValue, 8> NewArgs;
-  for (auto Arg : Apply.getArgumentsWithoutSelf()) {
-    NewArgs.push_back(Arg);
+  unsigned NumApplyArgs = Apply.getNumArguments();
+  bool UpdatedArgs = false;
+  for (unsigned ArgIdx = 0; ArgIdx < NumApplyArgs; ArgIdx++) {
+    llvm::SmallDenseMap<unsigned, const ConcreteExistentialInfo *, 8>::iterator
+        ArgIt = CEIs.find(ArgIdx);
+    if (ArgIt != CEIs.end()) {
```

Comment:
I think the == case should come first, with a 'continue' at the end. Then the rest does not need a whole level of indent.